server/sonar-server/src/main/java/org/sonar/server/computation/component/PathAwareVisitor.java

```diff
@@ -0,0 +1,295 @@
+  @Override
+  public void visit(Component component) {
+    if (component.getType().isDeeperThan(maxDepth)) {
+      return;
+    }
+
+    stack.add(new PathElementImpl<>(component, createForComponent(component)));
+
+    if (order == PRE_ORDER) {
+      visitNode(component);
+    }
+
+    visitChildren(component);
+
+    if (order == POST_ORDER) {
+      visitNode(component);
+    }
+
+    stack.pop();
+  }
+
+  private T createForComponent(Component component) {
+    switch (component.getType()) {
+      case PROJECT:
+        return factory.createForProject(component);
+      case MODULE:
+        return factory.createForModule(component);
+      case DIRECTORY:
+        return factory.createForDirectory(component);
+      case FILE:
+        return factory.createForFile(component);
+      default:
+        return factory.createForUnknown(component);
+    }
+  }
+
+  private void visitChildren(Component component) {
+    if (component.getType() != maxDepth) {
+      for (Component child : component.getChildren()) {
+        visit(child);
+      }
+    }
+  }
+
+  private void visitNode(Component component) {
+    visitAny(component, stack);
+    switch (component.getType()) {
+      case PROJECT:
+        visitProject(component, stack);
+        break;
+      case MODULE:
+        visitModule(component, stack);
+        break;
+      case DIRECTORY:
+        visitDirectory(component, stack);
+        break;
+      case FILE:
+        visitFile(component, stack);
+        break;
+      default:
+        visitUnknown(component, stack);
+    }
+  }
+
+  protected void visitProject(Component project, Path<T> path) {
+    // empty implementation, meant to be override at will by subclasses
+  }
+
+  protected void visitModule(Component module, Path<T> path) {
+    // empty implementation, meant to be override at will by subclasses
+  }
+
+  protected void visitDirectory(Component directory, Path<T> path) {
+    // empty implementation, meant to be override at will by subclasses
+  }
+
+  protected void visitFile(Component file, Path<T> path) {
+    // empty implementation, meant to be override at will by subclasses
+  }
+
+  protected void visitUnknown(Component unknownComponent, Path<T> path) {
+    // empty implementation, meant to be override at will by subclasses
+  }
+
+  protected void visitAny(Component component, Path<T> path) {
+    // empty implementation, meant to be override at will by subclasses
+  }
+
+  public interface StackElementFactory<T> {
+    T createForProject(Component project);
+
+    T createForModule(Component module);
+
+    T createForDirectory(Component directory);
+
+    T createForFile(Component file);
+
+    T createForUnknown(Component file);
+  }
+
+  /**
+   * A Simple implementation which uses the same factory method for all types which can be implemented by subclasses:
+   * {@link #createForAny(Component)}.
+   */
+  public static abstract class SimpleStackElementFactory<T> implements StackElementFactory<T> {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Reorder the modifiers to comply with the Java Language Specification. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AModifiersOrderCheck)